src/mesh/amr_loadbalance.cpp

```diff
@@ -45,6 +45,241 @@
 
 namespace parthenon {
 
+#ifdef MPI_PARALLEL
+//----------------------------------------------------------------------------------------
+//! \fn int CreateAMRMPITag(int lid, int ox1, int ox2, int ox3)
+//  \brief calculate an MPI tag for AMR block transfer
+// tag = local id of destination (remaining bits) + ox1(1 bit) + ox2(1 bit) + ox3(1 bit)
+//       + physics(5 bits)
+
+// See comments on BoundaryBase::CreateBvalsMPITag()
+
+int CreateAMRMPITag(int lid, int ox1, int ox2, int ox3) {
+  // the trailing zero is used as "id" to indicate an AMR related tag
+  return (lid << 8) | (ox1 << 7) | (ox2 << 6) | (ox3 << 5) | 0;
+}
+
+MPI_Request SendCoarseToFine(int lid_recv, int dest_rank, const LogicalLocation &fine_loc,
+                             Variable<Real> *var, Mesh *pmesh) {
+  MPI_Request req;
+  MPI_Comm comm = pmesh->GetMPIComm(var->label());
+  const int ox1 = ((fine_loc.lx1 & 1LL) == 1LL);
+  const int ox2 = ((fine_loc.lx2 & 1LL) == 1LL);
+  const int ox3 = ((fine_loc.lx3 & 1LL) == 1LL);
+  int tag = CreateAMRMPITag(lid_recv, ox1, ox2, ox3);
+  if (var->IsAllocated()) {
+    PARTHENON_MPI_CHECK(MPI_Isend(var->data.data(), var->data.size(), MPI_PARTHENON_REAL,
+                                  dest_rank, tag, comm, &req));
+  } else {
+    PARTHENON_MPI_CHECK(
+        MPI_Isend(var->data.data(), 0, MPI_PARTHENON_REAL, dest_rank, tag, comm, &req));
```

Comment:
Yeah, these are required. The receiver has to look for a message from every field, allocated or not, since it doesn't know the allocation status *a priori*. The zero size message still gets sent (i.e. it is not a null op), the receiver can probe for it and check its size, and if it is zero it knows the field was not allocated on the sending block.

Reply:
Thanks for the explanation. Makes sense. The "packing" message might end up on our top performance concerns sooner or later anyway.